src/Microsoft.ML.Data/Scorers/FeatureContributionCalculation.cs

```diff
@@ -0,0 +1,449 @@
+                _outputSchema = builder.GetSchema();
+                if (InputSchema[InputRoleMappedSchema.Feature.Index].HasSlotNames(InputRoleMappedSchema.Feature.Type.VectorSize))
+                    InputSchema[InputRoleMappedSchema.Feature.Index].Metadata.GetValue(MetadataUtils.Kinds.SlotNames, ref _slotNames);
+                else
+                    _slotNames = VBufferUtils.CreateEmpty<ReadOnlyMemory<char>>(InputRoleMappedSchema.Feature.Type.VectorSize);
+            }
+            else
+            {
+                _outputSchema = Schema.Create(new FeatureContributionSchema(_env, DefaultColumnNames.FeatureContributions,
+                    new VectorType(NumberType.R4, schema.Feature.Type as VectorType),
+                    InputSchema, InputRoleMappedSchema.Feature.Index));
+            }
+
+            _outputGenericSchema = _genericRowMapper.OutputSchema;
+            OutputSchema = new CompositeSchema(new Schema[] { _outputGenericSchema, _outputSchema, }).AsSchema;
+        }
+
+        /// <summary>
+        /// Returns the input columns needed for the requested output columns.
+        /// </summary>
+        public Func<int, bool> GetDependencies(Func<int, bool> predicate)
+        {
+            for (int i = 0; i < OutputSchema.Count; i++)
+            {
+                if (predicate(i))
+                    return col => col == InputRoleMappedSchema.Feature.Index;
+            }
+            return col => false;
+        }
+
+        public Row GetRow(Row input, Func<int, bool> active)
+        {
+            Contracts.AssertValue(input);
+            Contracts.AssertValue(active);
+            var totalColumnsCount = 1 + _outputGenericSchema.Count;
+            var getters = new Delegate[totalColumnsCount];
+
+            if (active(totalColumnsCount - 1))
+            {
+                getters[totalColumnsCount - 1] = _parent.Stringify
+                    ? _parent.GetTextContributionGetter(input, InputRoleMappedSchema.Feature.Index, _slotNames)
+                    : _parent.GetContributionGetter(input, InputRoleMappedSchema.Feature.Index);
+            }
+
+            var genericRow = _genericRowMapper.GetRow(input, GetGenericPredicate(active));
+            for (var i = 0; i < _outputGenericSchema.Count; i++)
+            {
+                if (genericRow.IsColumnActive(i))
+                    getters[i] = RowCursorUtils.GetGetterAsDelegate(genericRow, i);
+            }
+
+            return new SimpleRow(OutputSchema, genericRow, getters);
+        }
+
+        public Func<int, bool> GetGenericPredicate(Func<int, bool> predicate)
+        {
+            return col => predicate(col);
+        }
+
+        public IEnumerable<KeyValuePair<RoleMappedSchema.ColumnRole, string>> GetInputColumnRoles()
+        {
+            yield return RoleMappedSchema.ColumnRole.Feature.Bind(InputRoleMappedSchema.Feature.Name);
+        }
+    }
+
+    internal sealed class FeatureContributionSchema : ISchema
```

Comment:
>FeatureContributionSchema [](start = 26, length = 25)

Not sure if you're aware but we're attempting to deprecate usage of `ISchema`, i.e.,

[https://github.com/dotnet/machinelearning/blob/7faa46c41fcd1a214fac7bfbdc8a9d1adbcc8d41/src/Microsoft.ML.Core/Data/IDataView.cs#L12-L16](https://github.com/dotnet/machinelearning/blob/7faa46c41fcd1a214fac7bfbdc8a9d1adbcc8d41/src/Microsoft.ML.Core/Data/IDataView.cs#L12-L16)

Was the introduction of this class absolutely necessary?

Reply:
I believe so, this part of the code came from TLC.

---
In reply to: [243037751](https://github.com/dotnet/machinelearning/pull/1847#discussion_r243037751) [](ancestors = 243037751)